README.md
# node-prom-metrics

Sample NodeJS wrapper for Prometheus metrics

## Example

If one run the example `example/express-server.js` and go to `localhost:9001/metrics`, will see the following metrics exposed. They are a very basic example of total requests (counter) and request duration (histogram).

```text
# HELP http_requests_total Total HTTP requets processed
# TYPE http_requests_total counter
http_requests_total{route="GET",method="/metrics",status="200",hostname="my-hostname",foo="bar"} 3

# HELP http_duration_histogram_seconds HTTP request latency distribution
# TYPE http_duration_histogram_seconds histogram
http_duration_histogram_seconds_bucket{le="0.01",route="GET",method="/metrics",status="200",hostname="my-hostname",foo="bar"} 3
http_duration_histogram_seconds_bucket{le="0.05",route="GET",method="/metrics",status="200",hostname="my-hostname",foo="bar"} 3
http_duration_histogram_seconds_bucket{le="0.1",route="GET",method="/metrics",status="200",hostname="my-hostname",foo="bar"} 3
http_duration_histogram_seconds_bucket{le="0.5",route="GET",method="/metrics",status="200",hostname="my-hostname",foo="bar"} 3
http_duration_histogram_seconds_bucket{le="1",route="GET",method="/metrics",status="200",hostname="my-hostname",foo="bar"} 3
http_duration_histogram_seconds_bucket{le="5",route="GET",method="/metrics",status="200",hostname="my-hostname",foo="bar"} 3
http_duration_histogram_seconds_bucket{le="10",route="GET",method="/metrics",status="200",hostname="my-hostname",foo="bar"} 3
http_duration_histogram_seconds_bucket{le="+Inf",route="GET",method="/metrics",status="200",hostname="my-hostname",foo="bar"} 3
http_duration_histogram_seconds_sum{route="GET",method="/metrics",status="200",hostname="my-hostname",foo="bar"} 0.005
http_duration_histogram_seconds_count{route="GET",method="/metrics",status="200",hostname="my-hostname",foo="bar"} 3
```


CHANGELOG.md
## [1.0.1](https://github.com/Octadesk-Tech/node-prom-metrics/compare/v1.0.0...v1.0.1) (2021-10-11)


### Bug Fixes

* including NODE_AUTH_TOKEN ([02481df](https://github.com/Octadesk-Tech/node-prom-metrics/commit/02481df3dc36d826d20387d038182c26ec84d095))

# 1.0.0 (2021-10-11)


### Bug Fixes

* bump package.json version ([a2681bd](https://github.com/Octadesk-Tech/node-prom-metrics/commit/a2681bd364160308aade8f558ea1a7384cc794ff))
* creating .npmrc ([9bef943](https://github.com/Octadesk-Tech/node-prom-metrics/commit/9bef9439c212d0740f3fc11a5a2f6e52c1ebf0c5))
* npx semantic-release ([c15f717](https://github.com/Octadesk-Tech/node-prom-metrics/commit/c15f717cf0bd21ddc438551abe3a4b41aeb89440))
* package name ([981db08](https://github.com/Octadesk-Tech/node-prom-metrics/commit/981db08e77c8b9b141df75159d07afa164669092))


### Features

* Creating packages automatically ([46d6d5b](https://github.com/Octadesk-Tech/node-prom-metrics/commit/46d6d5bf82a616d4c46c46afc2b821d6ff79a1d3))


src/index.js
const Prometheus = require('prom-client')

class Metrics {
  constructor() {
    // Setup the register where all the metrics will be stored
    this._register = Prometheus.register

    // Enable default metrics gathering such as heap size, garbage collector stuff etc
    Prometheus.collectDefaultMetrics()

    // Set the content-type header reference for client retrieving
    this.contentType = this._register.contentType
  }

  getInstance() {
    if (!Metrics.instance) {
      Metrics.instance = new Metrics()
    }
    return Metrics.instance
  }

  /**
   * Set default labels to be appended to every single metric generated in the whole execution
   * based on an object {key1:value1, key2:value2}
   * @param {Object} labelNames
   */
  setDefaultLabelNames(labelNames) {
    this._register.setDefaultLabels(labelNames)
  }

  /**
   * Despite the fact that this class is totally server-framework independent, this method is
   * intended to help as a callback for route resolving in express.js scenarios (example) so
   * one can retrieve all the metrics stored in Prometheus registry
   *
   * app.get('/metrics', (req, res) => {
   *  res.set('Content-Type', Metrics.contentType)
   *  res.end(Metrics.metricsCallback())
   * });
   */
  metricsCallback() {
    return this._register.metrics()
  }

  /**
   * Return a metric from Prometheus registry to be used
   * @param {string} metricName
   */
  get(metricName) {
    return this._register.getSingleMetric(metricName)
  }

  /**
   * Extend Prometheus Counter class, which provides the following methods
   *  inc(number)         :void               Increment the counter by the provided number value
   *  labels(string[])    :Counter.Internal   Return the counter with the provided labels set
   *  reset()             :void               Reset the counter value to zero
   *  remove(string[])    :void               Remove metrics for the provided labels
   * @param {string}      metricName
   * @param {string}      metricName
   * @param {string[]}    labelNames
   */
  counter(name, help, labelNames) {
    let c = new Prometheus.Counter({
      name: name,
      help: help,
      labelNames: labelNames
    })
    this._register.registerMetric(c)
  }

  /**
   * Extend Prometheus Gauge class, which provides the following methods
   *  inc(number)         :void           Increment the counter by the provided number value
   *  dec(number)         :void           Increment the counter by the provided number value
   *  set(number)         :void           Set the gauge to the provided number value
   *  labels(string[])    :Gauge.Internal Return the gauge with the provided labels set
   *  reset()             :void           Reset the counter value to zero
   *  remove(string[])    :void           Remove metrics for the provided labels
   *
   * There is also a special case of this class named Timer. In fact this is a concept and not
   * another class itself
   *  startTimer()        :function       Starts a timer which will be the gauge value once you
   *                                      call the function returned on this method call
   * @param {string}      name
   * @param {string}      help
   * @param {string[]}    labelNames
   */
  gauge(name, help, labelNames) {
    let g = new Prometheus.Gauge({
      name: name,
      help: help,
      labelNames: labelNames
    })
    this._register.registerMetric(g)
  }

  /**
   * Extend Prometheus Histogram class, which provides the following methods
   * startTimer()     :function           Start a timer which will be the histogram value once
   *                                      you call the function returned on this method call
   * labels(string[]) :Histogram.Internal Return the histogram with the provided labels set
   * observe(number)  :voic               Observe value in histogram
   * reset()          :void               Reset the counter value to zero
   * remove(string[]) :void               Remove metrics for the provided labels
   * @param {string}      name
   * @param {string}      help
   * @param {string[]}    labelNames
   * @param {number[]}    buckets
   */
  histogram(name, help, labelNames, buckets) {
    let h = new Prometheus.Histogram({
      name: name,
      help: help,
      labelNames: labelNames,
      buckets: buckets
    })
    this._register.registerMetric(h)
  }

  /**
   * Extend Prometheus Summary class, which provides the following methods
   * startTimer()     :function           Start a timer which will be the summary value once
   *                                      you call the function returned on this method call
   * labels(string[]) :Summary.Internal   Return the summary with the provided labels set
   * observe(number)  :void               Observe value in summary
   * reset()          :void               Reset the counter value to zero
   * remove(string[]) :void               Remove metrics for the provided labels
   * @param {string}      name
   * @param {string}      help
   * @param {string[]}    labelNames
   * @param {number[]}    percentiles
   */
  summary(name, help, labelNames, percentiles) {
    let s = new Prometheus.Summary({
      name: name,
      help: help,
      labelNames: labelNames,
      percentiles: percentiles
    })
    this._register.registerMetric(s)
  }
}

module.exports = new Metrics()


example/express-server.js
const express = require("express")
const Metrics = require("../src").getInstance()
const morgan = require("morgan")
const os = require("os")

const PORT = 9001
const app = express()

app.use(morgan('combined'))

// Middleware function to count total requests and the request duration for a given method
app.use((req, res, next) => {
	res.locals.startTime = Date.now()

	next()
	
	Metrics.get("http_requests_total")
		.labels(req.method, req.route.path, res.statusCode)
		.inc()
	
	const responseTimeMS = Date.now() - res.locals.startTime
	Metrics.get("http_duration_histogram_seconds")
		.labels(req.method, req.route.path, res.statusCode)
	  	.observe(responseTimeMS/1000)
})

app.get('/', (req, res, next) => {
	res.send({
		message: "Hello world!"
	})
})

app.get('/metrics', (req, res, next) => {
	res.set('Content-Type', Metrics.contentType)
	res.end(Metrics.metricsCallback())
})

// Not found middleware
app.use((req, res, next) => {
	res.status(404).send({ 
		error: "Not found"
	})
})

// Error middleware
app.use((err, req, res, next) => {
	res.status(500).send({ 
		error: err 
	})
})

app.listen(PORT, () => {
	console.log(`Server listening to ${PORT}, metrics exposed on /metrics endpoint`)
	
	Metrics.setDefaultLabelNames({
		hostname: os.hostname(),
		foo: "bar"
	})

	// Register a requests_total metric to be used anywhere in the application
	Metrics.counter("http_requests_total", "Total HTTP requets processed", ["route", "method", "status"])
	
	// Register a http_duration metric to be used anywhere in the application
	Metrics.histogram("http_duration_histogram_seconds", "HTTP request latency distribution", 
		["route", "method", "status"], [0.01, 0.05, 0.1, 0.5, 1, 5, 10])
});